File: code/ch09/9.2.1.1.linked_list_cycle.py
class Solution:
    def hasCycle(self, head:ListNode)->bool:
        if head is None:
            return False

        node_set = {}
        while head.next:
            node_set[head] = True

            if head.next in node_set:
                return True

            head = head.next

        return False

File: code/ch09/9.2.1.1.linked_list_cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_set<ListNode *> seenNode;
        while (head) {
            if (seenNode.find(head) != seenNode.end())
                return true;
            seenNode.insert(head);
            head = head->next;
        }
        return false;  
    }
};
